sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/SaveAsHiveFile.scala

```diff
@@ -97,12 +99,46 @@ private[hive] trait SaveAsHiveFile extends DataWritingCommand {
       options = Map.empty)
   }
 
-  protected def getExternalTmpPath(
+  // Mostly copied from Context.java#getMRTmpPath of Hive 2.3.
+  // Visible for testing.
+  private[execution] def getNonBlobTmpPath(
+      hadoopConf: Configuration,
+      sessionScratchDir: String,
+      scratchDir: String): Path = {
+
+    // Hive's getMRTmpPath uses nonLocalScratchPath + '-mr-10000',
+    // which is ruled by 'hive.exec.scratchdir' including file system.
+    // This is the same as Spark's #oldVersionExternalTempPath.
+    // Only difference between #oldVersionExternalTempPath and Hive 2.3.0's is HIVE-7090.
+    // HIVE-7090 added user_name/session_id on top of 'hive.exec.scratchdir'
+    // Here it uses session_path unless it's emtpy, otherwise uses scratchDir.
+    val sessionPath = if (!sessionScratchDir.isEmpty) sessionScratchDir else scratchDir
+    val mrScratchDir = oldVersionExternalTempPath(new Path(sessionPath), hadoopConf, sessionPath)
+    logDebug(s"MR scratch dir '$mrScratchDir/-mr-10000' is used")
+    val path = new Path(mrScratchDir, "-mr-10000")
+    val scheme = Option(path.toUri.getScheme).getOrElse("")
+    if (scheme.equals("file")) {
+      logWarning(s"Temporary data will be written into a local file system " +
+        s"(scheme: '$scheme', path: '$mrScratchDir'). If your Spark is not in local mode, " +
+        s"you might need to configure 'hive.exec.scratchdir' " +
+        s"to use accessible file system (e.g. HDFS path) from any executors in the cluster.")
+    }
+    path
+  }
+
+  private def supportSchemeToUseNonBlobStore(path: Path): Boolean = {
+    path != null && {
+      val supportedBlobSchemes = SQLConf.get.supportedSchemesToUseNonBlobstore
+      val scheme = Option(path.toUri.getScheme).getOrElse("")
+      Utils.stringToSeq(supportedBlobSchemes).contains(scheme.toLowerCase(Locale.ROOT))
+    }
+  }
+
+  def getExternalTmpPath(
       sparkSession: SparkSession,
       hadoopConf: Configuration,
       path: Path): Path = {
     import org.apache.spark.sql.hive.client.hive._
-
```

Comment:
nit: please revert unnecessary change where possible.

Reply:
Thanks for pointing it. Reverted.